esp/services/esdl_svc_engine/esdl_svc_custom.cpp

```diff
@@ -0,0 +1,346 @@
+            cur.toDBGLog ();
+        }
+
+        ForEachItemIn(nci, m_childOtherwiseClauses)
+        {
+            CEsdlCustomTransformChoose & cur =
+            m_childOtherwiseClauses.item (nci);
+            cur.toDBGLog ();
+         }
+    }
+    DBGLOG (">>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>WHEN>>>>>>>>>>");
+}
+#endif
+
+//
+// CEsdlCustomTransform methods
+//
+CEsdlCustomTransform::CEsdlCustomTransform(IPropertyTree *cfg)
+{
+    m_xpathContext.set(getXpathContext());
+    Owned<IPropertyTreeIterator> transformsIter = cfg->getElements("xsdl:CustomRequestTransform");
+    {
+        ForEach(*transformsIter)
+        {
+            IPropertyTree & currentTransform = transformsIter->query();
+            m_name.set(currentTransform.queryProp("@name"));
+            //if m_ruleName.length() == 0?
+            DBGLOG("Compiling custom ESDL Transform: '%s'", m_name.str());
+
+            Owned<IPropertyTreeIterator> conditionalIterator = currentTransform.getElements("xsdl:choose");
+            ForEach(*conditionalIterator)
+            {
+                auto xslchooseelement = &conditionalIterator->query();
+                Owned<CEsdlCustomTransformChoose> currconditional = new CEsdlCustomTransformChoose(xslchooseelement);
+                m_customTransformClauses.append(*LINK(currconditional));
+            }
+        }
+    }
+#if defined(_DEBUG)
+    toDBGLog();
+#endif
+}
+
+void CEsdlCustomTransform::processTransform(IEspContext * context, StringBuffer & request, IPropertyTree * bindingCfg)
+{
+    if (request.length()!=0)
+    {
+#if defined(_DEBUG)
+        DBGLOG("ORIGINAL REQUEST: %s", request.str());
+
+        StringBuffer marshalled;
+        toXML(bindingCfg, marshalled.clear());
+        DBGLOG("INCOMING CONFIG: %s", marshalled.str());
+#endif
+
+        m_xpathContext->setXmlDoc(request.str());
+
+        VStringBuffer ver("%g", context->getClientVersion());
+        if(!m_xpathContext->addParameter("clientversion", ver.str()))
+            ERRLOG("Could not set custom transform variable: clientversion:'%s'", ver.str());
+
+        auto user = context->queryUser();
+        if (user)
+        {
+            Owned<IPropertyIterator> userPropIt = user->getPropertyIterator();
+            ForEach(*userPropIt)
+            {
+                const char *name = userPropIt->getPropKey();
+                if (name && *name)
+                    m_xpathContext->addParameter(name, user->getProperty(name));
+            }
+        }
+
+        Owned<IPropertyTreeIterator> configParams = bindingCfg->getElements("Transform/Param");
+        {
+            ForEach(*configParams)
+            {
+                IPropertyTree & currentParam = configParams->query();
+                m_xpathContext->addParameter(currentParam.queryProp("@name"), currentParam.queryProp("@value"));
+            }
+        }
+
+        Owned<IPropertyTree> thereq = createPTreeFromXMLString(request.str());
```

Comment:
Doing string->ptree->string conversions twice hurts the performance quite significantly. Any chance the transform can be done before esdltransformer? Or maybe add a form of esdl transform that will keep the transformed result in a ptree format (under the condition of no performance penalty)?

Reply:
Very valid comment regarding the representation and transformation of the request from str to ptree back to str. Delayed the transform as long as possible. At the point we receive the req for custom transform, it's already been processed by esdl engine and returned as STR. We transform (late) the req into a tree in order to perform custom transform.